### DISCUSSION TITLE
Understanding XCBBuildServiceProxyKit

### DISCUSSION THREAD

**Discussion by AlexDenisov** [MobileNativeFoundation/discussions#178] [General]
Hi folks, while trying to understand the inner workings of the `XCBBuildService` I discovered this fantastic tool.
I did look at the source code, and read the available documentation, but some pieces are still missing.
I'd highly appreciate if someone more knowledgable than I am can shed some light on it 😄

1. I'm certainly missing more high-level understanding of which role does `XCBBuildServiceProxyKit` play in the whole build process. E.g., when I run `xcodebuild` from a command line, who creates the proxy service? And who then instantiates the actual `XCBBuildService`? How are the messages transferred between these two? etc.
2. The other, smaller issue, is that I could not find which build commands the proxy can "intercept." I'm particularly interested in all the "builtin" commands (e.g., `builtin-copy`, `builtin-copyPlist`, etc.) and was wondering if this is something `XCBBuildServiceProxyKit` can help with.

Thank you!

**Comment by brentleyjones** [comment]
Hi Alex,

Improved documentation is a medium-term goal for the project. I'm currently focused on getting [rules_xcodeproj](https://github.com/buildbuddy-io/rules_xcodeproj) to 1.0, but post 1.0 there will be a need for that project to integrate with XCBBuildServiceProxyKit, at which point I plan on giving the project the love it needs to bring it up to production quality.

To answer some of your immediate questions though:
1. Nested in the [Updating Guide](https://github.com/MobileNativeFoundation/XCBBuildServiceProxyKit/blob/main/Docs/UPDATING.md) is a nice overview of how to install a proxy, with the main process being to replace the binary at `$(XCODE_DIR)/Contents/SharedFrameworks/XCBuild.framework/PlugIns/XCBBuildService.bundle/Contents/MacOS/XCBBuildService` and the secondary method being overriding the `XCBBUILDSERVICE_PATH` environment variable.
2. I don't know of a way to intercept the built-in commands, though I haven't tried to do that. Please let me know what you find!

**Reply by brentleyjones** [reply]
An update on this: we are now starting to integrate XCBBuildServiceProxyKit with rules_xcodeproj. As part of that I'm going to basically be rewriting the project with the knowledge gained over the last couple years, using many Swift language improvements (e.g. Swift concurrency), and greatly improving the documentation.
